zerver/actions/message_edit.py

```diff
@@ -1298,6 +1270,85 @@ def check_time_limit_for_change_all_propagate_mode(
     )
 
 
+def get_message_edit_request_object(
+    message: Message,
+    user_profile: UserProfile,
+    propagate_mode: str,
+    stream_id: int | None = None,
+    topic_name: str | None = None,
+    content: str | None = None,
+) -> StreamMessageEditRequest | DirectMessageEditRequest:
+    if not message.is_stream_message():
+        # We have already validated the code to have content
+        # as not None.
+        assert content is not None
+        return DirectMessageEditRequest(
+            content=content,
+            orig_content=message.content,
+            is_content_edited=True,
+        )
+
+    is_content_edited = False
+    new_content = message.content
+    if content is not None:
+        is_content_edited = True
+        if content.rstrip() == "":
+            content = "(deleted)"
+        new_content = normalize_body(content)
+
+    is_topic_edited = False
+    topic_resolved = False
+    topic_unresolved = False
+    old_topic_name = message.topic_name()
+    new_topic_name = old_topic_name
+    pre_truncation_new_topic_name = old_topic_name
+
+    if topic_name is not None:
+        is_topic_edited = True
+        pre_truncation_new_topic_name = topic_name
+        new_topic_name = truncate_topic(topic_name)
+
+        resolved_prefix_len = len(RESOLVED_TOPIC_PREFIX)
+        topic_resolved = (
+            new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+            and not old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+            and pre_truncation_new_topic_name[resolved_prefix_len:] == old_topic_name
+        )
+        topic_unresolved = (
+            old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+            and not new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+            and old_topic_name.lstrip(RESOLVED_TOPIC_PREFIX) == new_topic_name
+        )
+
+    orig_stream_id = message.recipient.type_id
+    orig_stream = get_stream_by_id_in_realm(orig_stream_id, message.realm)
+
+    is_stream_edited = False
+    new_stream = orig_stream
+    if stream_id is not None:
+        new_stream = access_stream_by_id_for_message(user_profile, stream_id, require_active=True)[
+            0
+        ]
+        is_stream_edited = True
+
+    return StreamMessageEditRequest(
+        is_content_edited=is_content_edited,
+        content=new_content,
+        is_topic_edited=is_topic_edited,
+        topic_name=new_topic_name,
+        pre_truncation_topic_name=pre_truncation_new_topic_name,
+        is_stream_edited=is_stream_edited,
+        topic_resolved=topic_resolved,
+        topic_unresolved=topic_unresolved,
+        orig_content=message.content,
+        orig_topic_name=old_topic_name,
+        orig_stream=orig_stream,
+        propagate_mode=propagate_mode,
+        stream=new_stream,
```

Comment:
I think `new_stream` or `target_stream` would be a better name in the dataclass; when we have two streams to talk about here, it's most readable if neither has bare `stream` as the name. Same with `topic_name`.

Reply:
`stream` field is set to the original stream if stream is not changed. This is essentially to avoid passing arguments like like `stream=message_edit_request.new_stream if message_edit_request.new_stream is not None else message_edit_request.orig_stream`. Similarly for topics as well.

So, `target_stream` and `target_topic_name` might be better name?

Reply:
I think `target_` seems clear, yeah.